3-enrich/stream-enrich/core/src/main/scala/com.snowplowanalytics.snowplow.enrich.stream/sources/Source.scala

```diff
@@ -158,35 +209,46 @@ abstract class Source(
    * @return Whether to checkpoint
    */
   def enrichAndStoreEvents(binaryData: List[Array[Byte]]): Boolean = {
-    val enrichedEvents = binaryData.flatMap(enrichEvents(_))
-    val successes = enrichedEvents collect { case Success(s) => s }
+    val enrichedEvents         = binaryData.flatMap(enrichEvents(_))
+    val successes              = enrichedEvents collect { case Success(s) => s }
     val sizeUnadjustedFailures = enrichedEvents collect { case Failure(s) => s }
     val failures = sizeUnadjustedFailures map {
-      case (value, key) => if (! isTooLarge(value)) {
-        value -> key
-      } else {
-        Source.adjustOversizedFailureJson(value) -> key
-      }
+      case (value, key) =>
+        if (!isTooLarge(value)) {
+          value -> key
+        } else {
+          Source.adjustOversizedFailureJson(value) -> key
+        }
     }
 
-    val (tooBigSuccesses, smallEnoughSuccesses) = successes partition { s => isTooLarge(s._1) }
+    val (tooBigSuccesses, smallEnoughSuccesses) = successes partition { s =>
+      isTooLarge(s._1)
+    }
 
     val sizeBasedFailures = for {
-      (value, key) <- tooBigSuccesses
-      m <- MaxRecordSize
+      (value, key, _) <- tooBigSuccesses
+      m               <- MaxRecordSize
     } yield Source.oversizedSuccessToFailure(value, m) -> key
 
-    val successesTriggeredFlush = goodSink.get.storeEnrichedEvents(smallEnoughSuccesses)
-    val failuresTriggeredFlush = badSink.get.storeEnrichedEvents(failures ++ sizeBasedFailures)
-    if (successesTriggeredFlush == true || failuresTriggeredFlush == true) {
+    val anonymizedSuccesses = smallEnoughSuccesses.map {
+      case (event, partition, _) => (event, partition)
+    }
+    val piiSuccesses = smallEnoughSuccesses.flatMap {
+      case (_, partition, pii) => pii.map((_, partition))
+    }
+
+    val successesTriggeredFlush = goodSink.get.storeEnrichedEvents(anonymizedSuccesses)
+    val piiTriggeredFlush       = piiSink.map(_.get.storeEnrichedEvents(piiSuccesses)).getOrElse(false)
```

Comment:
one thing I missed during the 1st review is, should we check for the sizes of the pii events separately from the usual enriched events?

Reply:
I am assuming that you are referring to the lines below `if(successesTriggeredFlush == true...`. Is that correct?

Reply:
Ok I understood now. So we can definitely check pii for size, but the question is, what to do if the are too big (they are still PII)?

Reply:
@alexanderdean Any suggestions?

Reply:
What are our options here @knservis ?

Reply:
I can see the following options:

* Do nothing: As this is using the *small enough* Enriched Events to emit PII we can assume that the PII event is going to be smaller than the parent event. This is a reasonable assumption IMHO. 
* Use the same limit and test PII separately which then creates the following two options for oversize PII events: 
* * If PII are bigger than the limit, throw them away (will create data gaps in piinguin)
* * Send them to the existing *bad* stream. That is problematic as it would leak PII to the bad stream, unless we somehow scrub the original values. Sounds like a much hassle for no actual benefit as there would still need to be a lot of manual work to recover.
* * Send the parent event _id and an appropriate message to *bad* (but if we are doing that, why not send it to PII and have alternative handling where we look up the parent event and do PII backfilling).
* Come up with an EnrichedEvent fragmentation protocol to handle all cases where the event is too large, where they are fragmented and reassembled. This may seem like a good option and it certainly will avoid dropping data, but it is problematic in that most (all?) projects operate on single events at the moment.

Reply:
Thanks for the detailed options @knservis ! Let's go with do nothing, but:

1. Create a ticket to flag that we are doing nothing, so we don't forget about this. Heavily link this new ticket back to this thread
2. Create a placeholder for exploring the fragmentation protocol idea, which is really neat. Again, heavily link it back to here

Thanks! @BenFradet shout if you disagree with the plan.

Reply:
Memo: https://github.com/snowplow/snowplow/issues/3755

Proposal: https://github.com/snowplow/snowplow/issues/3756
👍